README.md
# 🤫 PrivateStatus

![GitHub tag (with filter)](https://img.shields.io/github/v/tag/tinyoverflow/minecraft-privatestatus?style=flat-square&label=version)
![GitHub issues](https://img.shields.io/github/issues/tinyoverflow/minecraft-privatestatus?style=flat-square&label=issues)
![GitHub last commit (branch)](https://img.shields.io/github/last-commit/tinyoverflow/minecraft-privatestatus/main?style=flat-square&label=updated)
![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/tinyoverflow/minecraft-privatestatus/main?style=flat-square&label=codefactor)
![GitHub Workflow Status (with event)](https://img.shields.io/github/actions/workflow/status/tinyoverflow/minecraft-privatestatus/maven.yml?style=flat-square&label=build)
[![bStats Servers](https://img.shields.io/bstats/servers/19291?style=flat-square&label=servers)][bstats-link]

With PrivateStatus you can hide the online status of your server from unwanted players. This is especially worthwhile for private servers with a whitelist to prevent this. This plugin only works on Paper 1.20+ and is not compatible with Spigot. This is because this plugin relies on the so called `ServerListPingEvent` which can only be cancelled when using Paper.

## Configuration

The configuration is in `plugins/PrivateStatus/config.yml` and looks like this. Please only edit the `expiration-minutes` option. The `known-addresses` section is managed by the plugin itself. The option `expiration-minutes` defines how long an address should be kept in the history. The plugin automatically checks every 10 minutes if entries expired. If so, it removes them.

```yaml
# The amount of minutes after which addresses will expire. Minimum: 10.
expiration-minutes: 1440

# A list of already known addresses.
known-addresses: {}
```

## Video Example (YouTube)

[![PrivateStatus Demo](http://img.youtube.com/vi/aRo1AMhukKs/maxresdefault.jpg)](https://youtu.be/aRo1AMhukKs)

## More

- [View on GitHub][github-link]
- [View on Modrinth][modrinth-link]
- [View on Hangar][hangar-link]
- [View on bStats][bstats-link]

[github-link]: https://github.com/tinyoverflow/minecraft-privatestatus
[modrinth-link]: https://modrinth.com/plugin/privatestatus
[hangar-link]: https://hangar.papermc.io/tinyoverflow/PrivateStatus
[bstats-link]: https://bstats.org/plugin/bukkit/PrivateStatus/19291


CHANGELOG.md
**Changes**

 - Change default timeout to 1440 minutes (one day)

src/main/java/me/tinyoverflow/privatestatus/AddressRepository.java
package me.tinyoverflow.privatestatus;

import me.tinyoverflow.privatestatus.events.AddAddressEvent;
import org.bukkit.OfflinePlayer;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;

public class AddressRepository
{
    private final Logger logger;
    private final HashMap<InetAddress, LocalDateTime> storage = new HashMap<>();
    private final ZoneOffset zoneOffset;

    public AddressRepository(Logger logger)
    {
        this.logger = logger;
        this.zoneOffset = OffsetDateTime.now().getOffset();
    }

    /**
     * Loads known addresses from a {@code List<Map<String, Object>>}.
     *
     * @param data List of data.
     */
    public void fromList(List<Map<String, Object>> data)
    {
        for (Map<String, Object> section : data) {
            if (section.get("ip") == null || section.get("expiration") == null) {
                continue;
            }

            String address = (String) section.get("ip");
            long expirationTimestamp = ((Integer) section.get("expiration")).longValue();

            try {
                storage.put(
                        InetAddress.getByName(address),
                        LocalDateTime.ofEpochSecond(expirationTimestamp, 0, zoneOffset)
                );
            }
            catch (UnknownHostException e) {
                logger.warning("Unknown host found in config. Skipping: " + address);
            }
        }

        // Notify the user about the loading state.
        logger.info("Loaded " + storage.size() + " addresses from configuration.");
    }

    public List<Map<String, Object>> toList()
    {
        List<Map<String, Object>> addressList = new ArrayList<>();

        for (Map.Entry<InetAddress, LocalDateTime> entry : storage.entrySet()) {
            InetAddress inetAddress = entry.getKey();
            LocalDateTime localDateTime = entry.getValue();

            Map<String, Object> data = new HashMap<>();
            data.put("ip", inetAddress.getHostAddress());
            data.put("expiration", localDateTime.toEpochSecond(zoneOffset));

            addressList.add(data);
        }

        return addressList;
    }

    /**
     * Checks whether the given address is known.
     *
     * @param inetAddress The address to check for.
     * @return {@code true} if the address is known, {@code false} otherwise.
     */
    public boolean hasAddress(InetAddress inetAddress)
    {
        return storage.containsKey(inetAddress);
    }

    /**
     * Remembers a player and the address that belongs to it.
     *
     * @param inetAddress The address to associate with that player.
     */
    public void add(OfflinePlayer player, InetAddress inetAddress, LocalDateTime expiration)
    {
        AddAddressEvent event = new AddAddressEvent(player, inetAddress, expiration);
        if (event.callEvent()) {
            storage.put(inetAddress, expiration);
        }
    }

    @SuppressWarnings("unchecked")
    public Map<InetAddress, LocalDateTime> getAll()
    {
        return (Map<InetAddress, LocalDateTime>) storage.clone();
    }

    public void remove(InetAddress address)
    {
        storage.remove(address);
    }
}


src/main/java/me/tinyoverflow/privatestatus/events/AddAddressEvent.java
package me.tinyoverflow.privatestatus.events;

import org.bukkit.OfflinePlayer;
import org.bukkit.event.Cancellable;
import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.NotNull;

import java.net.InetAddress;
import java.time.LocalDateTime;

public class AddAddressEvent extends Event implements Cancellable
{
    private static final HandlerList HANDLER_LIST = new HandlerList();
    private final OfflinePlayer offlinePlayer;
    private final InetAddress address;
    private final LocalDateTime expiration;
    private boolean cancelled = false;

    public AddAddressEvent(OfflinePlayer offlinePlayer, InetAddress address, LocalDateTime expiration)
    {
        this.offlinePlayer = offlinePlayer;
        this.address = address;
        this.expiration = expiration;
    }

    @SuppressWarnings("unused")
    public static HandlerList getHandlerList()
    {
        return HANDLER_LIST;
    }

    @SuppressWarnings("unused")
    public OfflinePlayer getOfflinePlayer()
    {
        return offlinePlayer;
    }

    @SuppressWarnings("unused")
    public InetAddress getAddress()
    {
        return address;
    }

    @SuppressWarnings("unused")
    public LocalDateTime getExpirationDateTime() {
        return expiration;
    }

    @Override
    public boolean isCancelled()
    {
        return cancelled;
    }

    @Override
    public void setCancelled(boolean cancel)
    {
        cancelled = cancel;
    }

    @Override
    public @NotNull HandlerList getHandlers()
    {
        return HANDLER_LIST;
    }
}


src/main/java/me/tinyoverflow/privatestatus/events/ExpireAddressEvent.java
package me.tinyoverflow.privatestatus.events;

import org.bukkit.OfflinePlayer;
import org.bukkit.event.Cancellable;
import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.NotNull;

/**
 * This event will be fired when an address should be checked for expiration.
 * If this event gets cancelled, the address will not be expired.
 */
public class ExpireAddressEvent extends Event implements Cancellable
{
    private static final HandlerList HANDLER_LIST = new HandlerList();
    private final OfflinePlayer player;
    private boolean cancelled = false;

    public ExpireAddressEvent(OfflinePlayer player)
    {
        this.player = player;
    }

    @SuppressWarnings("unused")
    public static HandlerList getHandlerList()
    {
        return HANDLER_LIST;
    }

    @Override
    public @NotNull HandlerList getHandlers()
    {
        return HANDLER_LIST;
    }

    @Override
    public boolean isCancelled()
    {
        return cancelled;
    }

    @Override
    public void setCancelled(boolean cancel)
    {
        cancelled = cancel;
    }

    @SuppressWarnings("unused")
    public OfflinePlayer getPlayer()
    {
        return player;
    }
}


src/main/java/me/tinyoverflow/privatestatus/jobs/PruneExpiredAddressesJob.java
package me.tinyoverflow.privatestatus.jobs;

import me.tinyoverflow.privatestatus.AddressRepository;

import java.net.InetAddress;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.logging.Logger;

public class PruneExpiredAddressesJob implements Runnable
{
    private final Logger logger;
    private final AddressRepository repository;

    public PruneExpiredAddressesJob(Logger logger, AddressRepository repository)
    {
        this.logger = logger;
        this.repository = repository;
    }

    @Override
    public void run()
    {
        logger.fine("Pruning expired addresses...");
        int removedAddresses = 0;

        Map<InetAddress, LocalDateTime> addresses = repository.getAll();
        for (Map.Entry<InetAddress, LocalDateTime> entry : addresses.entrySet()) {
            InetAddress address = entry.getKey();
            LocalDateTime expiration = entry.getValue();

            if (expiration.isBefore(LocalDateTime.now())) {
                repository.remove(address);
                removedAddresses++;
            }
        }

        logger.fine("Pruning completed. Removed " + removedAddresses + " expired addresses.");
    }
}


src/main/java/me/tinyoverflow/privatestatus/PrivateStatus.java
package me.tinyoverflow.privatestatus;

import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
import me.tinyoverflow.privatestatus.jobs.PruneExpiredAddressesJob;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;

public class PrivateStatus extends JavaPlugin implements Listener
{
    private final String CONFIG_EXPIRATION_MINUTES = "expiration-minutes";
    private final String CONFIG_KNOWN_ADDRESSES = "known-addresses";

    private AddressRepository repository;

    @Override
    public void onLoad()
    {
        // Initialize repository.
        repository = new AddressRepository(getLogger());

        // Set sensible configuration defaults.
        getConfig().options().copyDefaults(true);

        getConfig().addDefault(CONFIG_EXPIRATION_MINUTES, 1440);
        getConfig().setComments(
                CONFIG_EXPIRATION_MINUTES,
                List.of("The amount of minutes after which addresses will expire. Minimum: 10.")
        );

        getConfig().addDefault(CONFIG_KNOWN_ADDRESSES, new HashMap<String, Object>());
        getConfig().setComments(CONFIG_KNOWN_ADDRESSES, List.of("A list of already known addresses."));

        saveConfig();
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onEnable()
    {
        getServer().getPluginManager().registerEvents(this, this);

        // Ensure the expiration minutes setting is at least 10.
        if (getConfig().getInt(CONFIG_EXPIRATION_MINUTES) < 10) {
            getConfig().set(CONFIG_EXPIRATION_MINUTES, 10);
        }

        // Load addresses from configuration file, if the section exists.
        List<?> data = getConfig().getList(CONFIG_KNOWN_ADDRESSES, new ArrayList<>());
        repository.fromList((List<Map<String, Object>>) data);
        //        ConfigurationSection configurationSection = getConfig().getConfigurationSection(CONFIG_KNOWN_ADDRESSES);
        //        if (configurationSection != null) {
        //            repository.fromMap(configurationSection.getValues(false));
        //        }

        // Initialize Metrics
        PrivateStatusMetrics metrics = new PrivateStatusMetrics(this, 19291);
        metrics.setExpirationMinutesMetric(getConfig().getInt(CONFIG_EXPIRATION_MINUTES));

        // Run prune job every 10 minutes.
        getServer().getScheduler().scheduleSyncRepeatingTask(
                this,
                new PruneExpiredAddressesJob(getLogger(), repository),
                0,
                10 * 60 * 20
        );
    }

    @Override
    public void onDisable()
    {
        // Save configuration to disk.
        getConfig().set(CONFIG_KNOWN_ADDRESSES, repository.toList());
        saveConfig();
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event)
    {
        Player player = event.getPlayer();

        InetSocketAddress socketAddress = player.getAddress();
        if (socketAddress == null) {
            getLogger().log(Level.WARNING, "Could not retrieve IP address of player " + player.getName());
            return;
        }

        InetAddress address = socketAddress.getAddress();
        if (address.isAnyLocalAddress()) {
            getLogger().log(Level.INFO, "Player " + player.getName() + " joined with local address. Ignoring.");
            return;
        }

        LocalDateTime expiration = LocalDateTime.now().plusMinutes(getConfig().getInt(CONFIG_EXPIRATION_MINUTES));
        repository.add(player, address, expiration);
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onServerListPingEvent(PaperServerListPingEvent event)
    {
        if (!repository.hasAddress(event.getAddress())) {
            event.setCancelled(true);
        }
    }
}
